### ISSUE TITLE
Key Error Issue

### ISSUE THREAD

**Issue by carynwillis** [theislab/cellrank#483]
I'm trying to run Cellrank on my own dataset following the tutorial. I can make it all the way to `cr.tl.terminal_states()` where I get this error:
```
Traceback (most recent call last):
  File "/app/cellrank/cellrank/tl/_init_term_states.py", line 99, in _initial_terminal
    kernel = PrecomputedKernel(key, adata=adata, backward=backward)
  File "/app/cellrank/cellrank/tl/kernels/_precomputed_kernel.py", line 59, in __init__
    transition_matrix = _read_graph_data(adata, transition_matrix)
  File "/app/cellrank/cellrank/ul/_utils.py", line 220, in _read_graph_data
    raise KeyError(f"Unable to find key `{key!r}` in `adata.obsp` or `adata.uns`.")
KeyError: "Unable to find key `'T_fwd'` in `adata.obsp` or `adata.uns`."

```
I have tried multiple fixes but can't figure out where the issue is coming from. Please advise. I included my python script. 

[cell_rank_issue.zip](https://github.com/theislab/cellrank/files/5914686/cell_rank_issue.zip)

**Comment by michalk8** [comment]
Hi @carynwillis ,

could you please tell me which version of CellRank you're using by running `import cellrank, print(cellrank.__version__`).
From what I can from the script you've provided, everything seems to be in order. In `cellrank/tl/_init_term_states.py`, we specifically catch the `KeyError` (which is used to determine whether to use an already computed transition matrix or not), see [here](https://github.com/theislab/cellrank/blob/master/cellrank/tl/_init_term_states.py#L101).

If the version is lower than `1.1.0`, I'd update it (this shouldn't really happen in `1.1.0`, though you might way a day, `1.2.0` is most likely coming out later today.

To fix this without updating, try running:
```python
cr.tl.transition_matrix(adata, weight_connectivities=0.2)
cr.tl.terminal_states(adata, cluster_key='leiden')
```

**Comment by carynwillis** [comment]
I am using version 1.1.0. I tried the fix above as well and got an error with the first line (cr.tl.transition_matrix(adata, weight_connectivities=0.2)): 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/cellrank/cellrank/tl/_transition_matrix.py", line 81, in transition_matrix
    **kwargs,
  File "/app/cellrank/cellrank/tl/kernels/_velocity_kernel.py", line 289, in compute_transition_matrix
    **kwargs,
  File "/app/cellrank/cellrank/ul/_utils.py", line 280, in wrapper
    return dispatch(args[0])(*args[1:], **kwargs)
  File "/app/cellrank/cellrank/tl/kernels/_velocity_kernel.py", line 571, in <lambda>
    lambda **kwargs: _run_in_parallel(_run_deterministic, **kwargs)
  File "/app/cellrank/cellrank/tl/kernels/_velocity_kernel.py", line 367, in _run_in_parallel
    )(**_filter_kwargs(fn, **kwargs))
  File "/app/cellrank/cellrank/ul/_parallelize.py", line 115, in wrapper
    return res if extractor is None else extractor(res)
  File "/app/cellrank/cellrank/tl/kernels/_velocity_kernel.py", line 364, in <lambda>
    extractor=lambda res: _reconstruct_one(np.concatenate(res, axis=-1), conn, ixs),
  File "/app/cellrank/cellrank/tl/kernels/_utils.py", line 207, in _reconstruct_one
    close_to_1 = np.isclose(probs.sum(1), 1.0)
  File "<__array_function__ internals>", line 6, in isclose
  File "/opt/conda/lib/python3.7/site-packages/numpy/core/numeric.py", line 2297, in isclose
    x = x * ones_like(cond)
  File "/opt/conda/lib/python3.7/site-packages/numpy/matrixlib/defmatrix.py", line 218, in __mul__
    return N.dot(self, asmatrix(other))
  File "<__array_function__ internals>", line 6, in dot
ValueError: shapes (17091,1) and (17091,1) not aligned: 1 (dim 1) != 17091 (dim 0)
```
I can try again tomorrow with 1.2.0 as well.

**Comment by michalk8** [comment]
@carynwillis 1.2.0 is up, https://github.com/theislab/cellrank/commit/cda144f75e9169b50170e59406e046cbe19fd1d4 should have fixed the
```pytb
ValueError: shapes (17091,1) and (17091,1) not aligned: 1 (dim 1) != 17091 (dim 0)
```

**Comment by carynwillis** [comment]
I have updated to 1.2.0 but I'm still getting errors when trying to run cr.tl.transition_matrix(adata, weight_connectivities=0.2):
`Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.7/multiprocessing/managers.py", line 597, in _run_server
    server.serve_forever()
  File "/opt/conda/lib/python3.7/multiprocessing/managers.py", line 173, in serve_forever
    sys.exit(0)
SystemExit: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()
  File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 133, in _remove_temp_dir
    rmtree(tempdir)
  File "/opt/conda/lib/python3.7/shutil.py", line 494, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/opt/conda/lib/python3.7/shutil.py", line 452, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/opt/conda/lib/python3.7/shutil.py", line 450, in _rmtree_safe_fd
    os.unlink(entry.name, dir_fd=topfd)
OSError: [Errno 16] Device or resource busy: '.nfs8059248500d0aaa50003447a'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/_transition_matrix.py", line 80, in transition_matrix
    **kwargs,
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/kernels/_velocity_kernel.py", line 294, in compute_transition_matrix
    **kwargs,
  File "/opt/conda/lib/python3.7/site-packages/cellrank/ul/_utils.py", line 279, in wrapper
    return dispatch(args[0])(*args[1:], **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/kernels/_velocity_kernel.py", line 576, in <lambda>
    lambda **kwargs: _run_in_parallel(_run_deterministic, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/kernels/_velocity_kernel.py", line 372, in _run_in_parallel
    )(**_filter_kwargs(fn, **kwargs))
  File "/opt/conda/lib/python3.7/site-packages/cellrank/ul/_parallelize.py", line 114, in wrapper
    return res if extractor is None else extractor(res)
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/kernels/_velocity_kernel.py", line 369, in <lambda>
    extractor=lambda res: _reconstruct_one(np.concatenate(res, axis=-1), conn, ixs),
  File "/opt/conda/lib/python3.7/site-packages/cellrank/tl/kernels/_utils.py", line 209, in _reconstruct_one
    f"Matrix is not row-stochastic. "
ValueError: Matrix is not row-stochastic. The following rows don't sum to 1: `[25, 32, 238, 323, 337, 366, 502, 571, 781, 1095, 1096, 1301, 1436, 1441, 1502, 1751, 2021, 2079, 2093, 2404, 2408, 2596, 2747, 2770, 2792, 3288, 3313, 3357, 3444, 3917, 3984, 4033, 4256, 4567, 4631, 4676, 4815, 4989, 5029, 5053, 5134, 5387, 5867, 5958, 6005, 6006, 6170, 6223, 6300, 6412, 6486, 6558, 6583, 6921, 7012, 7057, 7101, 7156, 7560, 7630, 7642, 7647, 7755, 7813, 7886, 8094, 8203, 8277, 8284, 8299, 8350, 8372, 8492, 8499, 8588, 8948, 9124, 9788, 9951, 9952, 9992, 10037, 10138, 10886, 10957, 11079, 11083, 11200, 11685, 11777, 11809, 11923, 11926, 12178, 12231, 12707, 12715, 12996, 13048, 13063, 13459, 13710, 13741, 13863, 13995, 14039, 14147, 14189, 14339, 14432, 14527, 14637, 14915, 15077, 15318, 15379, 15408, 15593, 15694, 15910, 16033, 16046, 16157, 16167, 16176, 16344, 16350]`

**Comment by michalk8** [comment]
I remember encountering this a while back (mostly because of numba, see [this comment](https://github.com/theislab/cellrank/blob/master/cellrank/tl/kernels/_utils.py#L192)), but I though I had fixed it but not using `numba` for the outermost parallelization.
I'd try these 3 things (in this order):
1. before running the program, disabling numba - this has to be done before importing cellrank as:
```python
import os
os.environ["NUMBA_DISABLE_JIT"] = "1"

# your imports go below
```
2. chaning the backend by running `cr.tl.transition_matrix(adata, weight_connectivities=0.2, backend='threading')`
3. if using `n_jobs=...` in `cellrank.tl.transition_matrix`, set it to `1`

If none of these work, I will make a branch where I try what's in the comment above, i.e. make an explicit copy of the matrix as
```
probs = csr_matrix((data[0].copy(), mat.indices, mat.indptr))
cors = csr_matrix((data[1].copy(), mat.indices, mat.indptr))
```

**Comment by Marius1311** [comment]
Hi @carynwillis , any success on this yet?

**Comment by carynwillis** [comment]
I have not had any success with the 3 steps above. I realized I did do Steps 1 and 2 together in the same run. Should they be done together or separate?

**Comment by michalk8** [comment]
> I have not had any success with the 3 steps above. I realized I did do Steps 1 and 2 together in the same run. Should they be done together or separate?

It could be done both ways, but since separately they didn't work, they are unlikely to work combined - you should try the fix I've implemented in #488 - hope that helps.

**Comment by Marius1311** [comment]
Maybe something specific about the transition matrix is going wrong here - @michalk8, don't we have some built in checks for connectedness of the graph and reducibility of the matrix? Could look into that.
